core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -128,15 +154,28 @@ public void clusterChanged(ClusterChangedEvent event) {
                     } else {
                         indexSettings = indexService.getIndexSettings();
                     }
-                    if (indicesService.canDeleteShardContent(shardId, indexSettings)) {
-                        deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                    IndicesService.ShardDeletionCheckResult shardDeletionCheckResult = indicesService.canDeleteShardContent(shardId, indexSettings);
+                    switch (shardDeletionCheckResult) {
+                        case CAN_DELETE:
+                            deleteShardIfExistElseWhere(event.state(), indexShardRoutingTable);
+                            break;
+                        case NO_FOLDER_FOUND:
+                        case SHARED_FILE_SYSTEM:
```

Comment:
It's not super fast. The IndexSettings object has to be created first just to do this check. I've seen cases where constructing this object again and again is non-negligible.